src/combinatorics/simple_graph/handshake.lean

```diff
@@ -0,0 +1,262 @@
+  thought of as being a directed edge.
+- `simple_graph.sum_degrees_eq_twice_card_edges` is the degree-sum formula.
+- `simple_graph.card_odd_degree_vertices_is_even` is the handshaking lemma.
+- `simple_graph.card_odd_degree_vertices_ne_is_odd` is that the number of odd-degree
+  vertices different from a given odd-degree vertex is odd.
+- `simple_graph.exists_ne_odd_degree_if_exists_odd` is that the existence of an
+  odd-degree vertex implies the existence of another one.
+
+## Implementation notes
+
+We give a combinatorial proof by using the fact that the map from
+darts to vertices has fibers whose cardinalities are the degrees and
+that the map from darts to edges is 2-to-1.
+
+## Tags
+
+simple graphs, sums, degree-sum formula, handshaking lemma
+-/
+open finset
+
+open_locale big_operators
+
+namespace simple_graph
+universes u
+variables {V : Type u} (G : simple_graph V)
+
+/-- A dart is a directed edge, consisting of an ordered pair of adjacent vertices. -/
+@[ext, derive decidable_eq]
+structure dart :=
+(fst snd : V)
+(is_adj : G.adj fst snd)
+
+/-- There is an equivalence between darts and pairs of a vertex and an incident edge. -/
+@[simps]
+def dart_equiv_sigma : G.dart ≃ Σ v, G.neighbor_set v :=
+{ to_fun := λ d, ⟨d.fst, d.snd, d.is_adj⟩,
+  inv_fun := λ s, ⟨s.fst, s.snd, s.snd.property⟩,
+  left_inv := λ d, by ext; simp,
+  right_inv := λ s, by ext; simp }
+
+instance dart.fintype [fintype V] [decidable_rel G.adj] : fintype G.dart :=
+fintype.of_equiv _ G.dart_equiv_sigma.symm
+
+instance dart.inhabited [inhabited V] [inhabited (G.neighbor_set (default _))] :
+  inhabited G.dart := ⟨G.dart_equiv_sigma.symm ⟨default _, default _⟩⟩
+
+variables {G}
+
+/-- The edge associated to the dart. -/
+def dart.edge (d : G.dart) : sym2 V := ⟦(d.fst, d.snd)⟧
+
+@[simp] lemma dart.edge_mem (d : G.dart) : d.edge ∈ G.edge_set :=
+d.is_adj
+
+/-- Reverses the orientation of a dart. -/
+def dart.rev (d : G.dart) : G.dart :=
+⟨d.snd, d.fst, G.sym d.is_adj⟩
+
+@[simp] lemma dart.rev_edge (d : G.dart) : d.rev.edge = d.edge :=
+sym2.eq_swap
+
+@[simp] lemma dart.rev_rev (d : G.dart) : d.rev.rev = d :=
+dart.ext _ _ rfl rfl
+
+@[simp] lemma dart_rev_involutive : function.involutive (dart.rev : G.dart → G.dart) :=
+dart.rev_rev
+
+lemma dart.rev_ne (d : G.dart) : d.rev ≠ d :=
+begin
+  cases d with f s h,
+  simp only [dart.rev, not_and, ne.def],
+  rintro rfl,
+  exact false.elim (G.loopless _ h),
+end
+
+lemma dart_edge_eq_iff (d₁ d₂ : G.dart) :
+  d₁.edge = d₂.edge ↔ d₁ = d₂ ∨ d₁ = d₂.rev :=
+begin
+  cases d₁ with s₁ t₁ h₁,
+  cases d₂ with s₂ t₂ h₂,
+  simp only [dart.edge, dart.rev_edge, dart.rev],
+  rw sym2.eq_iff,
+end
+
+variables (G)
+
+/-- For a given vertex `v`, the injective map from the incidence set at `v` to the darts there. --/
+def dart_from_neighbor_set (v : V) : G.neighbor_set v → G.dart :=
+λ w, ⟨v, w, w.property⟩
+
+lemma dart_from_neighbor_set_inj (v : V) : function.injective (G.dart_from_neighbor_set v) :=
```

Comment:
[See the last sentence of the naming conventions](https://leanprover-community.github.io/contribute/naming.html#injectivity).
```suggestion
lemma dart_of_neighbor_set_injective (v : V) : function.injective (G.dart_from_neighbor_set v) :=
```